The failure is:
```console
03:18 TASK [ansible-galaxy-collection : setup test collections for install and download test] ***
04:36 An exception occurred during task execution. To see the full traceback, use -vvv. The error was: tarfile.LinkOutsideDestinationError: 'docs/REÅDMÈ.md' would link to '/root/.ansible/tmp/ansible-moduletmp-1692920399.4914837-vvowxzjy/symlink-symlink-1.0.0/REÅDMÈ.md', which is outside the destination
04:36 fatal: [testhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 133, in <module>\n  File \"<stdin>\", line 125, in _ansiballz_main\n  File \"<stdin>\", line 73, in invoke_module\n  File \"<frozen runpy>\", line 226, in run_module\n  File \"<frozen runpy>\", line 98, in _run_module_code\n  File \"<frozen runpy>\", line 88, in _run_code\n  File \"/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py\", line 280, in <module>\n  File \"/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py\", line 276, in main\n  File \"/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py\", line 260, in run_module\n  File \"/usr/lib/python3.12/multiprocessing/pool.py\", line 774, in get\n    raise self._value\n  File \"/usr/lib/python3.12/multiprocessing/pool.py\", line 125, in worker\n    result = (True, func(*args, **kwds))\n                    ^^^^^^^^^^^^^^^^^^^\n  File \"/usr/lib/python3.12/multiprocessing/pool.py\", line 48, in mapstar\n    return list(map(*args))\n           ^^^^^^^^^^^^^^^^\n  File \"/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py\", line 161, in publish_collection\n  File \"/usr/lib/python3.12/tarfile.py\", line 2253, in extractall\n    tarinfo = self._get_extract_tarinfo(member, filter_function, path)\n              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/lib/python3.12/tarfile.py\", line 2307, in _get_extract_tarinfo\n    self._handle_fatal_error(e)\n  File \"/usr/lib/python3.12/tarfile.py\", line 2305, in _get_extract_tarinfo\n    tarinfo = filter_function(tarinfo, path)\n              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/lib/python3.12/tarfile.py\", line 820, in data_filter\n    new_attrs = _get_filtered_attrs(member, dest_path, True)\n                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/lib/python3.12/tarfile.py\", line 807, in _get_filtered_attrs\n    raise LinkOutsideDestinationError(member, target_path)\ntarfile.LinkOutsideDestinationError: 'docs/REÅDMÈ.md' would link to '/root/.ansible/tmp/ansible-moduletmp-1692920399.4914837-vvowxzjy/symlink-symlink-1.0.0/REÅDMÈ.md', which is outside the destination\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```
_(https://dev.azure.com/ansible/ansible/_build/results?buildId=87831&view=logs&j=70eeb688-fc21-5491-4420-b60b604f62ec&t=d6f17d19-0bb3-5e63-8427-e2ae23c55414&l=876)_

I extracted unescaped exception for readability:
```python-traceback
Traceback (most recent call last):
  File "<stdin>", line 133, in <module>
  File "<stdin>", line 125, in _ansiballz_main
  File "<stdin>", line 73, in invoke_module
  File "<frozen runpy>", line 226, in run_module
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py", line 280, in <module>
  File "/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py", line 276, in main
  File "/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py", line 260, in run_module
  File "/usr/lib/python3.12/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/usr/lib/python3.12/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
           ^^^^^^^^^^^^^^^^
  File "/tmp/ansible_setup_collections_payload_88i9xwj9/ansible_setup_collections_payload.zip/ansible/modules/setup_collections.py", line 161, in publish_collection
  File "/usr/lib/python3.12/tarfile.py", line 2253, in extractall
    tarinfo = self._get_extract_tarinfo(member, filter_function, path)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/tarfile.py", line 2307, in _get_extract_tarinfo
    self._handle_fatal_error(e)
  File "/usr/lib/python3.12/tarfile.py", line 2305, in _get_extract_tarinfo
    tarinfo = filter_function(tarinfo, path)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/tarfile.py", line 820, in data_filter
    new_attrs = _get_filtered_attrs(member, dest_path, True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/tarfile.py", line 807, in _get_filtered_attrs
    raise LinkOutsideDestinationError(member, target_path)
tarfile.LinkOutsideDestinationError: 'docs/REÅDMÈ.md' would link to '/root/.ansible/tmp/ansible-moduletmp-1692920399.4914837-vvowxzjy/symlink-symlink-1.0.0/REÅDMÈ.md', which is outside the destination

```

--- a/changelogs/fragments/tarfile_extract_warn.yml
+++ b/changelogs/fragments/tarfile_extract_warn.yml
@@ -1,3 +1,4 @@
 ---
-minor_changes:
+bugfixes:
 - tarfile - handle data filter deprecation warning message for extract and extractall (https://github.com/ansible/ansible/issues/80832).
+- ansible-galaxy - implemented probing mechanism for broken data filter in tarfile.

--- a/lib/ansible/galaxy/role.py
+++ b/lib/ansible/galaxy/role.py
@@ -391,14 +391,29 @@ class GalaxyRole(object):
                                         continue
                                     n_final_parts.append(n_part)
                                 member.name = os.path.join(*n_final_parts)
-                                maybe_filter_kwargs = {
-                                    'filter': 'data',
-                                } if hasattr(tarfile, 'data_filter') else {}
-                                # deprecated: description='extract fallback without filter' python_version='3.11'
-                                role_tar_file.extract(
-                                    member, to_native(self.path),
-                                    **maybe_filter_kwargs,
-                                )
+
+                                # Implemented the following code to circumvent broken implementation of data_filter
+                                # in tarfile. See for more information - https://github.com/python/cpython/issues/107845
+                                # deprecated: description='probing broken data filter implementation' python_version='3.12'
+                                HAS_FUNCTIONING_DATA_FILTER = False
+                                if hasattr(tarfile, 'data_filter'):
+                                    # We explicitly check if tarfile.data_filter is broken or not
+                                    ti = tarfile.TarInfo('docs/README.md')
+                                    ti.type = tarfile.SYMTYPE
+                                    ti.linkname = '../README.md'
+
+                                    try:
+                                        tarfile.data_filter(ti, '/foo')
+                                    except tarfile.LinkOutsideDestinationError:
+                                        pass
+                                    else:
+                                        HAS_FUNCTIONING_DATA_FILTER = True
+
+                                if HAS_FUNCTIONING_DATA_FILTER:
+                                    # deprecated: description='extract fallback without filter' python_version='3.11'
+                                    role_tar_file.extract(member, to_native(self.path), filter='data')  # type: ignore[call-arg]
+                                else:
+                                    role_tar_file.extract(member, to_native(self.path))
 
                         # write out the install info file for later use
                         self._write_galaxy_install_info()

--- a/lib/ansible/galaxy/role.py
+++ b/lib/ansible/galaxy/role.py
@@ -391,11 +391,14 @@ class GalaxyRole(object):
                                         continue
                                     n_final_parts.append(n_part)
                                 member.name = os.path.join(*n_final_parts)
+                                maybe_filter_kwargs = {
+                                    'filter': 'data',
+                                } if hasattr(tarfile, 'data_filter') else {}
                                 # deprecated: description='extract fallback without filter' python_version='3.11'
-                                if hasattr(tarfile, 'data_filter'):
-                                    role_tar_file.extract(member, to_native(self.path), filter='data')  # type: ignore[call-arg]
-                                else:
-                                    role_tar_file.extract(member, to_native(self.path))
+                                role_tar_file.extract(
+                                    member, to_native(self.path),
+                                    **maybe_filter_kwargs,
+                                )
 
                         # write out the install info file for later use
                         self._write_galaxy_install_info()

--- a/packaging/release.py
+++ b/packaging/release.py
@@ -1342,11 +1342,11 @@ def test_sdist() -> None:
             except FileNotFoundError:
                 raise ApplicationError(f"Missing sdist: {sdist_file.relative_to(CHECKOUT_DIR)}") from None
 
+            maybe_filter_kwargs = {
+                'filter': 'data',
+            } if hasattr(tarfile, 'data_filter') else {}
             # deprecated: description='extractall fallback without filter' python_version='3.11'
-            if hasattr(tarfile, 'data_filter'):
-                sdist.extractall(temp_dir, filter='data')  # type: ignore[call-arg]
-            else:
-                sdist.extractall(temp_dir)
+            sdist.extractall(temp_dir, **maybe_filter_kwargs)  # type: ignore[arg-type]
 
         pyc_glob = "*.pyc*"
         pyc_files = sorted(path.relative_to(temp_dir) for path in temp_dir.rglob(pyc_glob))

--- a/packaging/release.py
+++ b/packaging/release.py
@@ -1342,11 +1342,11 @@ def test_sdist() -> None:
             except FileNotFoundError:
                 raise ApplicationError(f"Missing sdist: {sdist_file.relative_to(CHECKOUT_DIR)}") from None
 
-            maybe_filter_kwargs = {
-                'filter': 'data',
-            } if hasattr(tarfile, 'data_filter') else {}
             # deprecated: description='extractall fallback without filter' python_version='3.11'
-            sdist.extractall(temp_dir, **maybe_filter_kwargs)  # type: ignore[arg-type]
+            if hasattr(tarfile, 'data_filter'):
+                sdist.extractall(temp_dir, filter='data')  # type: ignore[call-arg]
+            else:
+                sdist.extractall(temp_dir)
 
         pyc_glob = "*.pyc*"
         pyc_files = sorted(path.relative_to(temp_dir) for path in temp_dir.rglob(pyc_glob))

--- a/test/integration/targets/ansible-galaxy-collection/library/setup_collections.py
+++ b/test/integration/targets/ansible-galaxy-collection/library/setup_collections.py
@@ -136,6 +136,7 @@ def publish_collection(module, collection):
                 fd.write(b'data')
 
             os.symlink(b_target_file, os.path.join(b_collection_dir, b_target_file + b'-link'))
+            os.symlink(temp_fd.name, os.path.join(b_collection_dir, b_target_file + b'-outside-link'))
             os.symlink(os.path.join(b'..', b_target_file), os.path.join(b_collection_dir, b'docs', b_target_file))
             os.symlink(os.path.join(b_collection_dir, b_target_file),
                        os.path.join(b_collection_dir, b'plugins', b_target_file))
@@ -155,14 +156,11 @@ def publish_collection(module, collection):
 
             # Extract the tarfile to sign the MANIFEST.json
             with tarfile.open(collection_path, mode='r') as collection_tar:
-                maybe_filter_kwargs = {
-                    'filter': 'data',
-                } if hasattr(tarfile, 'data_filter') else {}
                 # deprecated: description='extractall fallback without filter' python_version='3.11'
-                collection_tar.extractall(
-                    path=os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version)),
-                    **maybe_filter_kwargs
-                )
+                if hasattr(tarfile, 'tar_filter'):
+                    collection_tar.extractall(path=os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version)), filter='tar')
+                else:
+                    collection_tar.extractall(path=os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version)))
 
             manifest_path = os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version), 'MANIFEST.json')
             signature_path = os.path.join(module.params['signature_dir'], '%s-%s-%s-MANIFEST.json.asc' % (namespace, name, version))

--- a/test/integration/targets/ansible-galaxy-collection/library/setup_collections.py
+++ b/test/integration/targets/ansible-galaxy-collection/library/setup_collections.py
@@ -136,7 +136,6 @@ def publish_collection(module, collection):
                 fd.write(b'data')
 
             os.symlink(b_target_file, os.path.join(b_collection_dir, b_target_file + b'-link'))
-            os.symlink(temp_fd.name, os.path.join(b_collection_dir, b_target_file + b'-outside-link'))
             os.symlink(os.path.join(b'..', b_target_file), os.path.join(b_collection_dir, b'docs', b_target_file))
             os.symlink(os.path.join(b_collection_dir, b_target_file),
                        os.path.join(b_collection_dir, b'plugins', b_target_file))

--- a/test/integration/targets/ansible-galaxy-collection/library/setup_collections.py
+++ b/test/integration/targets/ansible-galaxy-collection/library/setup_collections.py
@@ -156,11 +156,14 @@ def publish_collection(module, collection):
 
             # Extract the tarfile to sign the MANIFEST.json
             with tarfile.open(collection_path, mode='r') as collection_tar:
+                maybe_filter_kwargs = {
+                    'filter': 'data',
+                } if hasattr(tarfile, 'data_filter') else {}
                 # deprecated: description='extractall fallback without filter' python_version='3.11'
-                if hasattr(tarfile, 'data_filter'):
-                    collection_tar.extractall(path=os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version)), filter='data')
-                else:
-                    collection_tar.extractall(path=os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version)))
+                collection_tar.extractall(
+                    path=os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version)),
+                    **maybe_filter_kwargs
+                )
 
             manifest_path = os.path.join(collection_dir, '%s-%s-%s' % (namespace, name, version), 'MANIFEST.json')
             signature_path = os.path.join(module.params['signature_dir'], '%s-%s-%s-MANIFEST.json.asc' % (namespace, name, version))

--- a/test/lib/ansible_test/_internal/commands/sanity/validate_modules.py
+++ b/test/lib/ansible_test/_internal/commands/sanity/validate_modules.py
@@ -160,10 +160,11 @@ class ValidateModulesTest(SanitySingleVersion):
 
                 with tarfile.open(path) as file:
                     # deprecated: description='extractall fallback without filter' python_version='3.11'
-                    if hasattr(tarfile, 'data_filter'):
-                        file.extractall(temp_dir, filter='data')  # type: ignore[call-arg]
-                    else:
-                        file.extractall(temp_dir)
+                    maybe_filter_kwargs = {
+                        'filter': 'data',
+                    } if hasattr(tarfile, 'data_filter') else {}
+                    # deprecated: description='extractall fallback without filter' python_version='3.11'
+                    file.extractall(temp_dir, **maybe_filter_kwargs)  # type: ignore[arg-type]
 
                 cmd.extend([
                     '--original-plugins', temp_dir,

--- a/test/lib/ansible_test/_internal/commands/sanity/validate_modules.py
+++ b/test/lib/ansible_test/_internal/commands/sanity/validate_modules.py
@@ -160,11 +160,10 @@ class ValidateModulesTest(SanitySingleVersion):
 
                 with tarfile.open(path) as file:
                     # deprecated: description='extractall fallback without filter' python_version='3.11'
-                    maybe_filter_kwargs = {
-                        'filter': 'data',
-                    } if hasattr(tarfile, 'data_filter') else {}
-                    # deprecated: description='extractall fallback without filter' python_version='3.11'
-                    file.extractall(temp_dir, **maybe_filter_kwargs)  # type: ignore[arg-type]
+                    if hasattr(tarfile, 'tar_filter'):
+                        file.extractall(temp_dir, filter='tar')  # type: ignore[call-arg]
+                    else:
+                        file.extractall(temp_dir)
 
                 cmd.extend([
                     '--original-plugins', temp_dir,

